Extensions/webnowplaying.js

```diff
@@ -10,164 +10,350 @@
+				this.communicationRevision = "legacy";
 			}
 		}
 	}
 
-	function fireEvent(event) {
-		const m = event.data;
-		const n = m.indexOf(" ");
-		let type = n === -1 ? m : m.substring(0, n);
-		type = type.toUpperCase();
-		const info = m.substring(n + 1);
+	sendUpdate() {
+		if (!this._ws || this._ws.readyState !== WebSocket.OPEN) return;
+		switch (this.communicationRevision) {
+			case "legacy":
+				SendUpdateLegacy(this);
+				break;
+			case "1":
+				SendUpdateRev1(this);
+				break;
+			default:
+				break;
+		}
+	}
+}
 
+function OnMessageLegacy(self, message) {
+	// Quite lengthy functions because we optimistically update spicetifyInfo after receiving events.
+	try {
+		const [type, data] = message.toUpperCase().split(" ");
 		switch (type) {
-			case "PLAYPAUSE":
+			case "PLAYPAUSE": {
 				Spicetify.Player.togglePlay();
+				self.spicetifyInfo.State = self.spicetifyInfo.State === "PLAYING" ? "PAUSED" : "PLAYING";
 				break;
+			}
 			case "NEXT":
 				Spicetify.Player.next();
 				break;
 			case "PREVIOUS":
 				Spicetify.Player.back();
 				break;
-			case "SETPOSITION":
-				Spicetify.Player.seek(parseInt(info) * 1000);
+			case "SETPOSITION": {
+				// Example string: SetPosition 34:SetProgress 0,100890207715134:
+				const [, positionPercentage] = message.toUpperCase().split(":")[1].split("SETPROGRESS ");
+				Spicetify.Player.seek(parseFloat(positionPercentage.replace(",", ".")));
 				break;
+			}
 			case "SETVOLUME":
-				Spicetify.Player.setVolume(parseInt(info) / 100);
+				Spicetify.Player.setVolume(parseInt(data) / 100);
 				break;
-			case "REPEAT":
+			case "REPEAT": {
 				Spicetify.Player.toggleRepeat();
+				self.spicetifyInfo.Repeat = self.spicetifyInfo.Repeat === "NONE" ? "ALL" : self.spicetifyInfo.Repeat === "ALL" ? "ONE" : "NONE";
 				break;
-			case "SHUFFLE":
+			}
+			case "SHUFFLE": {
 				Spicetify.Player.toggleShuffle();
+				self.spicetifyInfo.Shuffle = !self.spicetifyInfo.Shuffle;
 				break;
-			case "RATING":
-				const like = parseInt(info) > 3;
-				const isLiked = storage.RATING > 3;
-				if ((like && !isLiked) || (!like && isLiked)) {
-					Spicetify.Player.toggleHeart();
-				}
+			}
+			case "TOGGLETHUMBSUP": {
+				Spicetify.Player.toggleHeart();
+				self.spicetifyInfo.Rating = self.spicetifyInfo.Rating === 5 ? 0 : 5;
 				break;
-			case "TOGGLETHUMBSUP":
-				if (!(storage.RATING > 3)) {
-					Spicetify.Player.toggleHeart();
-				}
+			}
+			// Spotify doesn't have a negative rating
+			// case 'TOGGLETHUMBSDOWN': break
+			case "RATING": {
+				const rating = parseInt(data);
+				const isLiked = storage.RATING > 3;
+				if (rating >= 3 && !isLiked) Spicetify.Player.toggleHeart();
+				else if (rating < 3 && isLiked) Spicetify.Player.toggleHeart();
+				self.spicetifyInfo.Rating = rating;
 				break;
-			case "TOGGLETHUMBSDOWN":
-				if (storage.RATING > 3) {
-					Spicetify.Player.toggleHeart();
-				}
```

Comment:
There is a `TOGGLETHUMBSDOWN` case here implemented previously as opposed to the comment you wrote.
Any particular reason it was removed or did you missed this part?

Reply:
Made it behave the same way as the web-ext does. If there is no "negative rating button" on the website, pressing "thumbs down" does nothing.

Reply:
Seems like the Heart/Remove Heart functionality has all been merged onto the thumbs up button, so it should be fine ig.